### PR TITLE
mgr/dashboard: Role management from the UI

### PR DESCRIPTION
**Pull request by ricardoasmarques**
This PR adds the role management feature into the UI.

screenshot from 2018-08-03 10-41-47

screenshot from 2018-08-03 10-42-01

### CODE REVIEW

**Comment by votdev** [comment]
What about implementing the system role column in the roles tab like this in RGW?
auswahl_009

**Comment by ricardoasmarques** [comment]
@votdev addressed your comments

File: src/pybind/mgr/dashboard/frontend/src/app/core/auth/role-form/role-form.component.html [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+          <label i18n
+                 class="control-label col-sm-3"
+                 for="name">Permissions
+          </label>
+          <div class="col-sm-9">
+            <table class="table table-bordered table-hover">
+              <thead>
+              <tr>
+                <th></th>
+                <th class="text-center">read</th>
```
Comment (votdev) [inline src/pybind/mgr/dashboard/frontend/src/app/core/auth/role-form/role-form.component.html]: Please make the column header text (Read, Create, Update, Delete) uppercase.

File: src/pybind/mgr/dashboard/frontend/src/app/core/auth/role-form/role-form.component.html [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+        <div class="form-group"
+             [ngClass]="{'has-error': roleForm.showError('description', formDir)}">
+          <label i18n
+                 class="control-label col-sm-3"
+                 for="name">Description
+          </label>
+          <div class="col-sm-9">
+            <input class="form-control"
+                   type="text"
+                   placeholder="Description..."
```
Comment (votdev) [inline src/pybind/mgr/dashboard/frontend/src/app/core/auth/role-form/role-form.component.html]: i18n-placeholder is missing

File: src/pybind/mgr/dashboard/frontend/src/app/core/auth/role-form/role-form.component.html [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+          <label i18n
+                 class="control-label col-sm-3"
+                 for="name">Name
+            <span class="required"
+                  *ngIf="mode !== roleFormMode.editing"></span>
+          </label>
+          <div class="col-sm-9">
+            <input class="form-control"
+                   type="text"
+                   placeholder="Name..."
```
Comment (votdev) [inline src/pybind/mgr/dashboard/frontend/src/app/core/auth/role-form/role-form.component.html]: i18n-placeholder is missing

**Comment by votdev** [comment]
Consider to use the CdValidator.unique in the role add formular to check if the role name already exists. See https://github.com/ceph/ceph/pull/23802 and https://github.com/votdev/ceph/blob/631a9c9844225ec47bb175a893da39795ec50e7b/src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-user-form/rgw-user-form.component.ts#L54 how to implement it.

File: src/pybind/mgr/dashboard/frontend/src/app/core/auth/role-form/role-form.component.ts [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+    private roleService: RoleService,
+    private scopeService: ScopeService,
+    private notificationService: NotificationService
+  ) {
+    this.createForm();
+  }
+
+  createForm() {
+    this.roleForm = new CdFormGroup({
+      name: new FormControl('', {
```
Comment (votdev) [inline src/pybind/mgr/dashboard/frontend/src/app/core/auth/role-form/role-form.component.ts]: Consider to use the CdValidators.unique in the role add formular to check if the role name already exists. See https://github.com/ceph/ceph/pull/23802 and https://github.com/votdev/ceph/blob/631a9c9844225ec47bb175a893da39795ec50e7b/src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-user-form/rgw-user-form.component.ts#L54 how to implement it.

File: src/pybind/mgr/dashboard/frontend/src/app/core/auth/role-form/role-form.component.ts [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+  }
+
+  editAction() {
+    const roleFormModel = this.getRequest();
+    this.roleService.update(roleFormModel).subscribe(
+      () => {
+        this.notificationService.show(
+          NotificationType.success,
+          `Role "${roleFormModel.name}" has been updated.`,
+          'Edit Role'
```
Comment (votdev) [inline src/pybind/mgr/dashboard/frontend/src/app/core/auth/role-form/role-form.component.ts]: Remove this line according to hacking.rst and https://github.com/ceph/ceph/blob/cb6a6967686e103d2a4bdffe6750a462eef826d2/src/pybind/mgr/dashboard/frontend/src/app/shared/services/notification.service.ts#L70.

File: src/pybind/mgr/dashboard/frontend/src/app/core/auth/role-form/role-form.component.ts [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+  }
+
+  createAction() {
+    const roleFormModel = this.getRequest();
+    this.roleService.create(roleFormModel).subscribe(
+      () => {
+        this.notificationService.show(
+          NotificationType.success,
+          `Role "${roleFormModel.name}" has been created.`,
+          'Create Role'
```
Comment (votdev) [inline src/pybind/mgr/dashboard/frontend/src/app/core/auth/role-form/role-form.component.ts]: Remove this line according to hacking.rst and https://github.com/ceph/ceph/blob/cb6a6967686e103d2a4bdffe6750a462eef826d2/src/pybind/mgr/dashboard/frontend/src/app/shared/services/notification.service.ts#L70.

File: src/pybind/mgr/dashboard/frontend/src/app/core/auth/role-form/role-form.component.ts [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+    return roleFormModel;
+  }
+
+  createAction() {
+    const roleFormModel = this.getRequest();
+    this.roleService.create(roleFormModel).subscribe(
+      () => {
+        this.notificationService.show(
+          NotificationType.success,
+          `Role "${roleFormModel.name}" has been created.`,
```
Comment (votdev) [inline src/pybind/mgr/dashboard/frontend/src/app/core/auth/role-form/role-form.component.ts]: Use ``Created role "${roleFormModel.name}"`` according to hacking.rst here.

File: src/pybind/mgr/dashboard/frontend/src/app/core/auth/role-form/role-form.component.ts [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+    );
+  }
+
+  editAction() {
+    const roleFormModel = this.getRequest();
+    this.roleService.update(roleFormModel).subscribe(
+      () => {
+        this.notificationService.show(
+          NotificationType.success,
+          `Role "${roleFormModel.name}" has been updated.`,
```
Comment (votdev) [inline src/pybind/mgr/dashboard/frontend/src/app/core/auth/role-form/role-form.component.ts]: Use ``Updated role "${roleFormModel.name}"`` according to hacking.rst here.

**Comment by ricardoasmarques** [comment]
@votdev @tspmelo I've addressed your comments

File: src/pybind/mgr/dashboard/frontend/src/app/core/auth/role-list/role-list.component.ts [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+  updateSelection(selection: CdTableSelection) {
+    this.selection = selection;
+  }
+
+  deleteRole(role: string) {
+    this.roleService.delete(role).subscribe(
+      () => {
+        this.getRoles();
+        this.modalRef.hide();
+        this.notificationService.show(NotificationType.success, `Deleted Role '${role}'`);
```
Comment (votdev) [inline src/pybind/mgr/dashboard/frontend/src/app/core/auth/role-list/role-list.component.ts]: I'm not sure how we handle this, but if we use correct english it should be ``Deleted role ...``. Maybe another topic for our style guide.

File: src/pybind/mgr/dashboard/frontend/src/app/core/auth/role-list/role-list.component.ts [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+      }
+    );
+  }
+
+  deleteRoleModal() {
+    this.modalRef = this.modalService.show(DeletionModalComponent);
+    const name = this.selection.first().name;
+    this.modalRef.content.setUp({
+      metaType: 'Role',
+      pattern: `${name}`,
```
Comment (votdev) [inline src/pybind/mgr/dashboard/frontend/src/app/core/auth/role-list/role-list.component.ts]: This is obsolete if PR #23897 is merged before yours.

File: src/pybind/mgr/dashboard/frontend/src/app/shared/api/role.service.ts [hunk 11]
```diff
@@ -12,4 +16,31 @@ export class RoleService {
...
+  create(role: RoleFormModel) {
+    return this.http.post(`api/role`, role);
+  }
+
+  update(role: RoleFormModel) {
+    return this.http.put(`api/role/${role.name}`, role);
+  }
+
+  exists(name: string): Observable<boolean> {
+    return this.list().pipe(
```
Comment (votdev) [inline src/pybind/mgr/dashboard/frontend/src/app/shared/api/role.service.ts]: Please add at least one unit test for this service method.

File: src/pybind/mgr/dashboard/frontend/src/app/core/auth/role-form/role-form-mode.enum.ts [hunk 12]
```diff
@@ -0,0 +1,3 @@
+export enum RoleFormMode {
+  editing = 'editing'
```
[the remark this answers is not in the document]
Comment (ricardoasmarques) [reply src/pybind/mgr/dashboard/frontend/src/app/core/auth/role-form/role-form-mode.enum.ts]: For now, I'll keep it as `string` for consistency with `user-form.*`

**Comment by ricardoasmarques** [comment]
@votdev I've addressed your comments

**Comment by votdev** [comment]
Naming the menu 'Access management' is another option.

**Comment by LenzGr** [comment]
When editing an existing user, I entered a few characters in the first password field and then removed them again, as I did not want to change the existing password. However, I was not able to proceed, as the dialogue insisted that the "Password confirmation doesn't match the password." - this does not happen if I make the same change to the "Confirm password" change.

**Comment by ricardoasmarques** [comment]
> When editing an existing user, I entered a few characters in the first password field and then removed them again, as I did not want to change the existing password. However, I was not able to proceed, as the dialogue insisted that the "Password confirmation doesn't match the password." - this does not happen if I make the same change to the "Confirm password" change.

This PR is for Role management only, if we have an issue in user management I suggest to fix it in a separate issue / PR.

**Comment by LenzGr** [comment]
> Naming the menu 'Access management' is another option.

It might, but that is a bit too vague in my opinion. Also, I noticed that the cog icon currently does not provide any tooltip when I hover above it with the mouse pointer. It should display "Dashboard Settings" (but that probably deserves a separate PR).

**Comment by LenzGr** [comment]
> This PR is for Role management only, if we have an issue in user management I suggest to fix it in a separate issue / PR.

Good point; will submit a separate issue for this.

**Comment by ricardoasmarques** [comment]
> A minor nit: would it be possible to change the toplevel menu entry from "Users" to "Users and Roles" or "Users / Roles"?

@LenzGr We will have more tabs here (e.g., the SSO settings), so I agree with @votdev that a more generic menu entry is a better approach (instead of adding all tab names to the toplevel menu item).

Also, now that I've revisited this topic, I no longer think that "Roles" should be a child of "Users", and probably it doesn't make sense to have `Administration` in the breadcrumb.

My new proposal is the following:

**Users**
screenshot from 2018-09-05 11-27-07

**Roles**
screenshot from 2018-09-05 11-27-23

@LenzGr @votdev does this works for you?

**Comment by LenzGr** [comment]
> @LenzGr @votdev does this works for you?

This works for me. Thank you!

**Comment by ricardoasmarques** [comment]
@LenzGr @votdev I've updated this PR with the breadcrumb improvement discussed in the previous comment.

**Pull request merged**